validator_client/src/block_service.rs

```diff
@@ -335,81 +335,235 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
             )
         }
 
-        for validator_pubkey in proposers {
-            let builder_proposals = self
-                .validator_store
-                .get_builder_proposals(&validator_pubkey);
-            let service = self.clone();
-            let log = log.clone();
-            self.inner.context.executor.spawn(
-                async move {
-                    if builder_proposals {
+        let deneb_fork_activated = self
+            .context
+            .eth2_config
+            .spec
+            .deneb_fork_epoch
+            .and_then(|fork_epoch| {
+                let current_epoch = self.slot_clock.now()?.epoch(E::slots_per_epoch());
+                Some(current_epoch >= fork_epoch)
+            })
+            .unwrap_or(false);
+
+        if self.validator_store.produce_block_v3() || deneb_fork_activated {
+            for validator_pubkey in proposers {
+                let service: BlockService<T, E> = self.clone();
+                let log = log.clone();
+                self.inner.context.executor.spawn(
+                    async move {
                         let result = service
                             .clone()
```

Comment:
This `.clone` can be removed I think?

Reply:
When removing I end up getting `borrowed data escapes outside of method` compiler errors. I think we need to pass an owned instance of `log` into the spawned thread to keep the compiler happy

Reply:
Removed here b93f16d282edac636739252dd7af6111a2c3e2d6